### ISSUE TITLE
Should warn when safe functions dereference raw pointer arguments

### ISSUE THREAD

**Issue by rillian** [rust-lang/rust#29257]
Not sure if this should be an RFC, but I think a warning would work if the present behaviour is as desired.

I'm working on C interface code. People have told me, 'Everything `extern "C"` should be unsafe.' I presume this is because C apis inevitably involve raw pointers. But that's not quite true. Otherwise the compiler would require it.

Consider these two cases:
```rust
fn foo(p: *const u8) {
  println!("ptr {:?}", p);
}

fn bar(p: *const u8) {
  match p.is_null() {
    true => println!("null ptr {:?}", p),
    false => unsafe { println!("non-null ptr {:?}: {:?}", p, *p) },
  }
}
```
https://play.rust-lang.org/?gist=90a9dc04959cb831751b&version=stable

`foo` is safe, because it never dereferences the pointer, and this can be a useful part of the interface, so we have a counterexample of a function taking a raw pointer which doesn't need to be declared unsafe.

`bar` might *look* safe, but it's not. It will behave as indended for valid pointers, like you'd get from `slice::as_ptr`. It will also safely handle null pointers like you might get from a C api. But it's perfectly possible to pass an *invalid* pointer, crashing or (worse) yielding garbage. For example, `bar(1 as *const u8)` in the play example above.

So `bar` **should** be declared unsafe, but the compiler lets me declare it safe instead. Which means people will get this wrong (as I did) and random library code will have rare crashes. This is not the promise of rust.

A warning would address the possibility of programmer error here, assuming I'm correct in what the unsafe keyword means in this context. Functions which are declared safe, but take raw pointer arguments, should warn if they dereference that pointer. There needs to be something to tie the interior unsafe block to the function signature.

**Comment by nagisa** [comment]
This is an interesting abuse of unsafe and a bad API. I think it could be linted against (see rust-clippy cc @Manishearth), but doing something about it in the compiler it is a nontrivial decision and is unlikely to happen.

AFAIR, the promise of `unsafe` is “when crashed due to memory safety issue, code which caused the issue is inside unsafe block” which isn’t technically incorrect in this case, but might be relevant when such broken function comes from an external crate.

Just seeing a safe function taking a raw pointer as parameter would ring a 150dB alarm in my head.

**Comment by sfackler** [comment]
An unsafe block is the programmer asserting that "I know what I'm doing" - at that point it's out of the compiler's hands. The unsafe block is there as a warning to readers that the code inside of it should be more carefully reviewed than normal, where, presumably, that `bar` is a broken function.

There is already a lint that checks for unsafe blocks: http://is.gd/cKQSpq

**Comment by Manishearth** [comment]
I agree with @sfackler. `unsafe` is supposed to be used when you're telling the compiler that the code _is_ safe, even if it can't verify it for you.

Verifying if code in `unsafe` blocks is actually safe is outside of the scope of rustc and in general an unsolvable problem.

**Comment by Manishearth** [comment]
> AFAIR, the promise of unsafe is “when crashed due to memory safety issue, code which caused the issue is inside unsafe block”

Not really. `unsafe` can end up infecting the entire module nearby, see the `capacity` field of `Vec`, for instance.

**Comment by nagisa** [comment]
@Manishearth code in an `unsafe` is a reason for failure nevertheless. Regardless of who it ends up infecting.

**Comment by Manishearth** [comment]
Yes. I'm saying that the code with the issue can be somewhere else.

**Comment by rillian** [comment]
> You don't have to scope unsafe so that it encloses everything which can cause an issue.

So are you disagreeing that bar() should be declared unsafe to signal the need for care with invalid pointers to the caller? I'm still really fuzzy on what this keyword means in practice.

**Comment by Manishearth** [comment]
I agree that it should be declared `unsafe`. I disagree that it's the compiler's job to do anything about it if it isn't.

**Comment by sfackler** [comment]
@rillian An unsafe function states that the caller is responsible for ensuring that some set of conditions are true to avoid undefined behavior/crashes/etc.

An unsafe block  in a safe function, on the other hand, states that the *author* is responsible for ensuring the overall function is actually safe. In the case of `bar`, that's obviously not true.

**Comment by rillian** [comment]
Ok, thanks for confirming. So, again, I think a warning is the correct thing to do here. Any invariant which isn't tested is liable to break. The compiler is in the best position to verify this particular one. In the C world, helping programmers notice code which is valid but can lead to correctness issues is what warnings are for.

I'm surprised by the hostility to this idea. There seems to be a feeling in the rust community that 'unsafe' means "you're on your own". I'd like to write only safe rust all the time too, but being a systems language means that's not always possible, and it doesn't help me solve my problems or feel welcome in the community. If you find yourself thinking this way, please reconsider.

Do you have a more concrete argument against adding a warning? An example where there would be false positives, for example?

**Comment by sfackler** [comment]
The warning already exists, it's called `unsafe_code`. Dereferencing a raw pointer is no more intrinsically sketchy than calling into an `extern "C"` function or calling an `unsafe` Rust function. Why should it be singled out with a separate warning?

**Comment by Manishearth** [comment]
> An example where there would be false positives, for example?

Basically, all of them. Most of the use cases where you _would_ use unsafe involve sprinklings of raw pointers.

In other words, such a warning would almost be like a warning for "hey, you're using unsafe". Which makes it pretty useless, since you're already opting in to something when you use `unsafe`; you shouldn't have to opt in to more or less the same thing with an `allow()` attribute. It would feel like the compiler is going "Are you sure? Ok, you are. But are you really, _really_ sure?".

> There seems to be a feeling in the rust community that 'unsafe' means "you're on your own".

The moment you start using unsafe, it's pretty clear that there are other complex external factors in play at the moment. At that point, it's hard to use static analysis to figure out anything special about the code. Even just warning on an unconditional dereference of a raw pointer is dubious -- plenty of cases when the raw pointer is known to be allocated (of course this reasoning only works for private functions).

**Comment by Manishearth** [comment]
Sorry if you feel we're being hostile -- we're trying to be helpful. We just don't see any incremental benefit of this over what already exists.

**Comment by petrochenkov** [comment]
>An example where there would be false positives, for example?

Something nonsensical like this:
```
fn f(mut ptr: *const u8) {
  let a = 10u8;
  ptr = &a;
  let b = unsafe { *ptr }; // this dereference is actually safe
}
```
I.e. the argument is replaced and only then used.

**Comment by sfackler** [comment]
More concretely, the implementations of `Vec`, `Box`, `Arc`, `Rc`, `HashMap`, `LinkedList`, etc all expose safe APIs that use and dereference raw pointers internally.

**Comment by petrochenkov** [comment]
It seems strange to me too, that this idea has no support, maybe people misunderstand your point?
Basically every public (!) function taking raw pointers as arguments and going to dereference them have a contract - the input pointer is valid. But it can't *rely* on this contract if it's safe, so it can never dereference this pointer!
Private functions can rely on some extra knowledge about invariants, because unsafety is module level, as was mentioned, so they can dereference the pointer based on this knowledge.

IMO this would be a very useful warning if applied to immutable arguments on public functions with near 100% error-detection rate.

**Comment by Manishearth** [comment]
Ah, right, if we scope this to public functions we're in business.

I'm okay with such a lint, then, since most raw pointer derefs occur in private code with module-scoped unsafety.

However, note that at the moment Rust isn't looking to add more lints, according to @alexcrichton. I'd be happy to accept this as a lint in rust-clippy, though.

**Comment by dimbleby** [comment]
cf #27456.  Also crabtw/rust-bindgen#231

**Comment by Manishearth** [comment]
Filed https://github.com/Manishearth/rust-clippy/issues/411

**Comment by Gankro** [comment]
For further information on the meaning of `unsafe`, see https://doc.rust-lang.org/nightly/nomicon/meet-safe-and-unsafe.html

**Comment by rillian** [comment]
Limiting this to public functions would work for the C api use case.

**Comment by Aatch** [comment]
Updated the title to reflect the *actual* thing being proposed here. Simply "warning when dereferencing raw pointers in a safe function" is not at all useful, warning that a safe function takes a raw pointer and dereferences *that* makes more sense.

**Comment by steveklabnik** [comment]
Since new lints have a big impact on users of rustc, the policy is that they should go through the RFC process like other user-facing changes. As such, I'm going to give this one a close, but if anyone comes across this ticket and wants this lint, consider adding it to clippy and/or writing up an RFC. Thanks!
